pygeo/parameterization/DVGeoCST.py

```diff
@@ -0,0 +1,1175 @@
+        dtype : type, optional
+            Type for instantiated arrays, by default float
+
+        Returns
+        -------
+        ndarray (# pts,)
+            Derivative of the y coordinates with respect to the first class shape parameter
+        """
+        C = DVGeometryCST.computeClassShape(x[x != 0.0], N1, N2, dtype=dtype)
+        S = DVGeometryCST.computeShapeFunctions(x[x != 0.0], w, dtype=dtype)
+        dydN1 = np.zeros_like(x, dtype=dtype)
+        dydN1[x != 0.0] = np.sum(S, axis=0) * C * np.log(x[x != 0.0])
+        return dydN1
+
+    @staticmethod
+    def computeCSTdydN2(x, N1, N2, w, dtype=float):
+        r"""Compute the derivatives of the height of a CST curve with respect to N2
+
+        Given :math:`y = C(x, N1, N2) * S(x)`
+
+        :math:`\frac{dy}{dN2} = S(x) * \frac{dC}{dN2} = S(x) * C(x, N1, N2) * \ln(1-x)`
+
+        This function assumes x has been normalised to the range [0,1].
+
+        Parameters
+        ----------
+        x : ndarray (# pts,)
+            x coordinates at which to compute the CST curve height
+        N1 : float
+            First class shape parameter
+        N2 : float
+            Second class shape parameter
+        w : ndarray (# coeff,)
+            CST coefficient array
+        dtype : type, optional
+            Type for instantiated arrays, by default float
+
+        Returns
+        -------
+        ndarray (# pts,)
+            Derivative of the y coordinates with respect to the second class shape parameter
+        """
+        C = DVGeometryCST.computeClassShape(x[x != 1.0], N1, N2, dtype=dtype)
+        S = DVGeometryCST.computeShapeFunctions(x[x != 1.0], w, dtype=dtype)
+        dydN2 = np.zeros_like(x, dtype=dtype)
+        dydN2[x != 1.0] = np.sum(S, axis=0) * C * np.log(1 - x[x != 1.0])
+        return dydN2
+
+    @staticmethod
+    def computeCSTfromCoords(xCoord, yCoord, nCST, N1=0.5, N2=1.0, dtype=float):
+        """
+        Compute the CST coefficients that fit a set of airfoil
+        coordinates (either for the upper or lower surface, not both).
+
+        This function internally normalizes the x and y-coordinates.
+
+        Parameters
+        ----------
+        xCoord : ndarray
+            Upper or lower surface airfoil x-coordinates (same length
+            as yCoord vector).
+        yCoord : ndarray
+            Upper or lower surface airfoil y-coordinates (same length
+            as xCoord vector).
+        nCST : int
+            Number of CST coefficients to fit.
+        N1 : float, optional
+            First class shape parameter to assume in fitting, by default 0.5
+        N2 : float, optional
+            Second class shape parameter to assume in fitting, by default 1.0
+        dtype : type, optional
+            Type for instantiated arrays, by default float
+
+        Returns
+        -------
+        np.ndarray (nCST,)
+            CST coefficients fit to the airfoil surface.
+        """
+        # Normalize x and y
```

Comment:
currently this fit is done in parallel on all procs. this is probably fine, but then we do have cases where the fit is ill conditioned, and it can likely lead to different fits on different procs. maybe its better to run this fit only on root and bcast w in the comm to other procs